### ISSUE TITLE
"No tests found" when using separate directory and config for eg e2e tests

### ISSUE THREAD

**Issue by aMOPel** [nvim-neotest/neotest-jest#61]
**Problem:**

I can't figure out how to configure this.

**Plugin Setup:**

All dependencies installed.
```
'nvim-neotest/neotest',
        dependencies = {
            'nvim-lua/plenary.nvim',
            'nvim-treesitter/nvim-treesitter',
            'antoinemadec/FixCursorHold.nvim',
            'haydenmeade/neotest-jest',
```

Setup Call
```
    require('neotest').setup {
      adapters = {
        require('neotest-jest')({
          jestCommand = "npx jest",
          jestConfigFile = "package.json",
          cwd = function(path)
            return vim.fn.getcwd()
          end,
        }),
      },
    }
```

**Project Setup:**

**Unit Tests:**

Jest config is in `package.json`
```json
{
  "jest": {
    "moduleFileExtensions": [
      "js",
      "json",
      "ts"
    ],
    "rootDir": "src",
    "testRegex": ".*\\.spec\\.ts$",
    "transform": {
      "^.+\\.(t|j)s$": "ts-jest"
    },
    "collectCoverageFrom": [
      "**/*.(t|j)s"
    ],
    "coverageDirectory": "../coverage",
    "testEnvironment": "node",
    "moduleNameMapper": {
      "src/(.*)$": "<rootDir>/$1"
    }
  }
}
```

Test files are in `src/*.spec.ts` 

Normal test cases work, eg `<Cmd>lua require('neotest').run.run()<CR>` in `src/app.controller.spec.ts`

**e2e tests:**

Jest config is in `test/jest-e2e.json`

```json
{
  "moduleFileExtensions": [
    "js",
    "json",
    "ts"
  ],
  "rootDir": ".",
  "testEnvironment": "node",
  "testRegex": ".e2e-spec.ts$",
  "transform": {
    "^.+\\.(t|j)s$": "ts-jest"
  },
  "moduleNameMapper": {
    "src/(.*)$": "<rootDir>/../src/$1"
  }
}
```

Test files are in `test/app/app.e2e-spec.ts`

Running tests with cli works: 
`NODE_ENV=test jest --config ./test/jest-e2e.json`
Running tests with neotest doesn't work:
`:lua require('neotest').run.run({env={NODE_ENV='test'}, jestConfigFile='./test/jest-e2e.json'})<cr>` in `test/app/app.e2e-spec.ts`
produces "No tests found".

Thank you for your work!

**Comment by CallumHoward** [comment]
I'm also seeing "No tests found" but I think this might be a recent regression, this was working fine just a couple of weeks ago!

**Comment by CallumHoward** [comment]
Yep, as I thought, if I pin the version to this commit: `3853b842b5d4850a2d5495f75802dabb3cda246c` then it works fine. Curious to know if that helps you

**Comment by HugoArts** [comment]
Probably a duplicate of #60. The problem is commit 0c0dad7 updated the detection logic so that `jest` has to appear in the dependencies/devDependencies of the package.json, otherwise it will refuse to detect any tests.

**Comment by aMOPel** [comment]
> Curious to know if that helps you

```
~/.local/share/nvim/site/pack/packer/start/neotest-jest $ git status
HEAD detached at 3853b84
nothing to commit, working tree clean
```

`:lua require('neotest').run.run({env={NODE_ENV='test'}, jestConfigFile='./test/jest-e2e.json'})<cr>` in `test/app/app.e2e-spec.ts`
still produces "No tests found".

**Comment by stephenmckinney** [comment]
I had to take it back a few commits before the dependency logic was added. Try commit `219033658bf8de6b91e201036a5cc5456905d9d2`. In my case my package.json looks like"

```json
  "devDependencies": {
...
    "@types/jest": "^27.4.1",
...
  },
```

**Comment by stephenmckinney** [comment]
@haydenmeade what's the rationale for locking this down so tightly?

**Comment by tolik505** [comment]
I have the same issue. It just ignores .e2e-spec.ts but works fine with .spec.ts

**Comment by tolik505** [comment]
I've added a related PR https://github.com/nvim-neotest/neotest-jest/pull/92

**Comment by MisanthropicBit** [comment]
@aMOPel Thanks for submitting an issue 🙂

The `getJestConfig` option can also accept a function that takes the current file path as argument. You could use that to see if you were running a normal test or an e2e test and return the right config file path.

The call `:lua require('neotest').run.run({env={NODE_ENV='test'}, jestConfigFile='./test/jest-e2e.json'})<cr>` won't work but `neotest-jest` now supports `neotest`'s `extra_args` field so calling `:lua require('neotest').run.run({env={NODE_ENV='test'}, extra_args = { "--config='./test/jest-e2e.json' } })<cr>` should work if jest prioritises the last `--config` option given on the command line. Not sure if you can pass environment variables like that but jest [automatically sets](https://jestjs.io/docs/environment-variables) `NODE_ENV` to `'test'` if not set to something else.

@stephenmckinney The rationale is that multiple test runners recognise the `test.ts` extension (e.g. `vitest`) so matching on the test file extension alone isn't enough to determine if an adapter should be activated or not. Checking if the jest config is given directly in the `package.json` file might be a sensible extension.

@tolik505 You can now pass a `isTestFile` option which can be a function to allow overriding the logic for matching test extensions. See [here](https://github.com/nvim-neotest/neotest-jest?tab=readme-ov-file#istestfile).

**Comment by MisanthropicBit** [comment]
Closing due to inactivity. Feel free to reopen.
